matrix-e2e.c

```diff
@@ -296,6 +298,161 @@ static int matrix_restore_e2e_account(MatrixConnectionData *conn)
+{
+  // TODO
+}
+
+/*
+ * Get a set of device keys for ourselves.  Either by retreiving it from our store
+ * or by generating a new set.
+ *
+ * Returns: 0 on success
+ */
+int matrix_e2e_get_device_keys(MatrixConnectionData *conn, const gchar *device_id)
+{
+    PurpleConnection *pc = conn->pc;
+    JsonObject * json_dev_keys = NULL;
+    OlmAccount *account = olm_account(g_malloc0(olm_account_size()));
+    char *pickled_account = NULL;
+    void *random_pot = NULL;
+    int ret = 0;
+
+    if (!conn->e2e) {
+        conn->e2e = g_new0(MatrixE2EData,1);
+        conn->e2e->device_id = g_strdup(device_id);
+    }
+    conn->e2e->oa = account;
+
+    /* Try and restore olm account from settings; may fail, may work
+     * or may say there were no settings stored.
+     */
+    ret = matrix_restore_e2e_account(conn);
+    purple_debug_info("matrixprpl",
+                      "restore_e2e_account says %d\n", ret);
+    if (ret < 0) {
+        goto out;
+    }
+
+    if (ret == 0) {
+        /* No stored account - create one */
+        size_t needed_random = olm_create_account_random_length(account);
+        random_pot = get_random(needed_random);
+        if (!random_pot) {
+            purple_connection_error_reason(pc,
+                    PURPLE_CONNECTION_ERROR_OTHER_ERROR,
+                    "Unable to get randomness");
+            ret = -1;
+            goto out;
+        };
+
+        if (olm_create_account(account, random_pot, needed_random) ==
+            olm_error()) {
+            purple_connection_error_reason(pc,
+                    PURPLE_CONNECTION_ERROR_OTHER_ERROR,
+                    olm_account_last_error(account));
+            ret = -1;
+            goto out;
+        }
+        ret = matrix_store_e2e_account(conn);
+        if (ret) {
+            goto out;
+        }
+    }
+
+    /* Form a device keys object for an upload,
+     * from https://matrix.org/speculator/spec/drafts%2Fe2e/client_server/unstable.html#post-matrix-client-unstable-keys-upload
+     */
+    json_dev_keys = json_object_new();
+    json_object_set_string_member(json_dev_keys, "user_id", conn->user_id);
+    json_object_set_string_member(json_dev_keys, "device_id", device_id);
+    /* Add 'algorithms' array - is there a way to get libolm to tell us the list of what's supported */
+    /* the output of olm_account_identity_keys isn't quite right for it */
+    JsonArray *algorithms = json_array_new();
+    json_array_add_string_element(algorithms, "m.olm.curve25519-aes-sha256");
+    json_array_add_string_element(algorithms, "m.megolm.v1.aes-sha");
+    json_object_set_array_member(json_dev_keys, "algorithms", algorithms);
+
+    /* Add 'keys' entry */
+    JsonObject *json_keys = json_object_new();
+    gchar **algorithm_strings, **key_strings;
+    int num_algorithms = get_id_keys(pc, account, &algorithm_strings,
+                                     &key_strings);
+    if (num_algorithms < 1) {
+        json_object_unref(json_dev_keys);
```

Comment:
this looks like it should be `json_keys` rather than `json_dev_keys`?

Reply:
oops, good catch.